### PR TITLE
[DIF] Introduce DIF-specific Style Guide

### PR DESCRIPTION
**Pull request by lenary**
This adds specific C style guidance for DIF libraries and DIFs. These
points address specific questions raised in early DIF reviews.

---

This is based on the consensus (consensi?) that we are close to reaching in the [DIF Implementation Questions Jan 2020](https://docs.google.com/document/d/1cLyNo5wNTx-ITZ0Cz9xI6L-3tKT5WTU_sYGwrUJuCz8/edit) Document. 

It does not yet address the structuring of DIF error code enums. 

This does not replace the overall "DIF Specification" document that is still in-progress.

### CODE REVIEW

File: sw/device/lib/dif/README.md [hunk 1]
```diff
@@ -12,3 +12,54 @@ libraries.
...
+  out-parameter for returning the value.
+
+* DIFs must check their arguments against preconditions. DIFs must "early
+  return" with an error code if their arguments fail the following checks:
+  * DIFs must ensure their pointer arguments are non-null, unless that pointer
+    is for an optional out-parameter.
+  * DIFs must ensure, if they only accept a subset of an enum, that the argument
+    is within that subset. However, DIFs may assume that the argument is one of
+    the enum variants.
+  * DIFs must not have side-effects before an "early return". Side-effects
```
Comment (asb) [inline sw/device/lib/dif/README.md]: Given that there are naturally going to be error (early return cases) other than precondition checks that would have to involve side effects, maybe rephrase. "DIFs must now have side effects before an "early return" due to a precondition check failure."

File: sw/device/lib/dif/README.md [hunk 2]
```diff
@@ -12,3 +12,54 @@ libraries.
...
+    include (but are not limited to) writing to memory, including memory-mapped
+    hardware, and modifying processor CSRs.
+
+* Switch statements in DIFs must always have a default case, including when
```
Comment (asb) [inline sw/device/lib/dif/README.md]: Unless you also enable `-Wswitch-enum` I think you're slightly weakening the compilers ability to warn for coding errors, in cases where you intend but fail to cover the whole enum.

I suspect enabling `-Wswitch-enum` would be too noisy, and so I'm not convinced this guidance is the right choice.

File: sw/device/lib/dif/README.md (hunk 1 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: I intended "early return" to be a clearer explanation of the "guard" construct mentioned in the document. It is not clearer, as "early return" is a slightly different concept, so I will rephrase this in terms of guard statements.

File: sw/device/lib/dif/README.md (hunk 2 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: The intention of `default` is to catch two cases:
- when a programmer has passed a representable value that does not correspond to an enum variant value.
- when someone has glitched the core, and a valid value was passed, but the hardware memory now contains an invalid but representable value.

`-Wswitch-enum` will give errors even if there is a `default`, but does not check for reachability (so if one variant is unreachable due to a prior guard, it will still error if there is no explicit `case` to the variant). I think there could be a case for enabling `-Wswitch-enum` for the DIFs, but I would like to see more DIF code before we do so. 

Personally, I would like to see as much help as possible for programmers who intended, but have not managed, to cover the whole enum. This I would like to see balanced with being prudent about glitching and invalid variants.

**Comment by lenary** [comment]
I will squash this commit when it is ready - but I do see the value of retaining the edit history while the topics are still under review.

File: sw/device/lib/dif/README.md [hunk 3]
```diff
@@ -12,3 +12,54 @@ libraries.
 There is one DIF library per hardware IP, and each one contains the DIFs
```
Comment (silvestrst) [inline sw/device/lib/dif/README.md]: Could you please for the first occurence of acronym `DIF` use full name. Like `Device Interface Functions (DIF)`.

File: sw/device/lib/dif/README.md [hunk 4]
```diff
@@ -12,3 +12,54 @@ libraries.
...
+  functionality may require an environment set up using another DIF library, but
+  DIFs must not call DIFs in other DIF libraries.
+
+### DIF Guidance
+
+* DIF declarations must match their definitions exactly.
+  * Scalar arguments should not be marked `const` in DIF signatures.
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can error or produce a value, it should return an enum value, and use an
```
Comment (silvestrst) [inline sw/device/lib/dif/README.md]: `If a DIF can error or produce a value...` and instead of or?

File: sw/device/lib/dif/README.md (hunk 3 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: There is one at the top of the file.

File: sw/device/lib/dif/README.md (hunk 4 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: I will rephrase to "If a DIF can either error or produce a value" to clarify that we should not be returning an error value and also using the out-parameter (I think the out-parameter should only be written to if the function succeeds, but I am probably overlooking a moment where this might be useful).

File: sw/device/lib/dif/README.md [hunk 5]
```diff
@@ -12,3 +12,56 @@ libraries.
...
+* DIF libraries can only depend on the headers and libraries in the
+  `sw/device/lib/base` directory.
+* DIF libraries must not depend on other DIF libraries. Exercising DIF
+  functionality may require an environment set up using another DIF library, but
+  DIFs must not call DIFs in other DIF libraries.
+
+### DIF Guidance
+
+* DIF declarations must match their definitions exactly.
+  * Scalar arguments should not be marked `const` in DIF signatures.
```
Comment (mcy) [inline sw/device/lib/dif/README.md]: I would prefer to say "cv-qualified" to catch `volatile` as well but that's up to you.

File: sw/device/lib/dif/README.md [hunk 6]
```diff
@@ -12,3 +12,56 @@ libraries.
...
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can either error or produce a value, it should return an enum return
+  code, and use an out-parameter for returning the produced value.
```
Comment (mcy) [inline sw/device/lib/dif/README.md]: This does not address what an infaillable DIF function should return (which is a potential thing that could exist). I'm in favor of
```
typedef enum dif_infaillable {
  kDifInfaillableOk,
} dif_infaillable_t;
```

File: sw/device/lib/dif/README.md [hunk 7]
```diff
@@ -12,3 +12,56 @@ libraries.
...
+
+* DIFs must check their arguments against preconditions using "guard
+  statements". A guard statement is a simple if statement at the start of a
+  function which only returns an error code if the preconditions are not met.
+  Guard statements must cover the following checks:
+  * DIFs must ensure their pointer arguments are non-null, unless that pointer
+    is for an optional out-parameter.
```
Comment (mcy) [inline sw/device/lib/dif/README.md]: Do we want to allow optional out-parameters? I think that's too much of a hassle.

File: sw/device/lib/dif/README.md [hunk 8]
```diff
@@ -12,3 +12,56 @@ libraries.
...
+  * DIFs must ensure, if they only accept a subset of an enum, that the argument
+    is within that subset. However, DIFs may assume that the argument is one of
+    the enum variants.
```
Comment (mcy) [inline sw/device/lib/dif/README.md]: Something something "DIFs may assume that passing an undefined enum value is instant UB"? I'm typically in favor of phrasing such illegal conditions as "instant Undefined Behavior", since it communicates the type of care the programmer should take.

File: sw/device/lib/dif/README.md [hunk 9]
```diff
@@ -12,3 +12,56 @@ libraries.
...
+
+* DIFs must use enum return codes rather than booleans for reporting errors. If
+  a DIF can either error or produce a value, it should return an enum return
+  code, and use an out-parameter for returning the produced value.
+
+* DIFs must check their arguments against preconditions using "guard
+  statements". A guard statement is a simple if statement at the start of a
+  function which only returns an error code if the preconditions are not met.
+  Guard statements must cover the following checks:
+  * DIFs must ensure their pointer arguments are non-null, unless that pointer
```
Comment (mcy) [inline sw/device/lib/dif/README.md]: I think this should clarify that `mmio_region_t` is not a pointer parameter that can be meaningfully checked for nullness. DIFs should never ever access the pointer inside, even for a null check, since that pointer even existing is an implementation detail (which tests, for example, may violate).

File: sw/device/lib/dif/README.md [hunk 10]
```diff
@@ -12,3 +12,56 @@ libraries.
...
+
+* Switch statements in DIFs must always have a default case, including when
+  switching on an enum value (an "enum switch").
+  * The default case of an enum switch should report an error with values that
+    are not an enum variant.
+  * Enum switches do not need a `case` for enum variants that are unreachable
+    due to a guard statement.
+
+* DIFs must use `sw/device/lib/mmio.h` for accessing memory-mapped hardware.
+  DIFs must not use `sw/device/lib/memory.h` for accessing memory-mapped hardware.
```
Comment (mcy) [inline sw/device/lib/dif/README.md]: Something something "Since this would require casting away `volatile`, this is already considered Undefined Behavior by C."

File: sw/device/lib/dif/README.md (hunk 6 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: Infallible DIFs should return a value (if there is one to return), or `void` if there is not. Infallible DIFs should not use out-parameters if they return `void`. I don't see any value in returning `dif_infaillable_t` where `void` would suffice.

File: sw/device/lib/dif/README.md (hunk 9 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: `mmio_region_t` is not a pointer parameter. It is a struct parameter. Yes, I will include that you should not be looking inside a `mmio_region_t`.

**Comment by alphan** [comment]
Do we have a recommendation/rule regarding using `inline` for `static` functions in `.c` files? (came up in #1254, @silvestrst).

**Comment by lenary** [comment]
`static` and `inline static` mean effectively the same thing in C11. I think we decided to err towards using only `static` to have the DIFs be a bit cleaner, and in case the DIFs are used as example code anywhere.

File: sw/device/lib/dif/README.md (hunk 6 above)
Comment (gkelly) [reply sw/device/lib/dif/README.md]: > Infallible DIFs should return a value (if there is one to return), or `void` if there is not. Infallible DIFs should not use out-parameters if they return `void`. I don't see any value in returning `dif_infaillable_t` where `void` would suffice.

Strongly agree.  It behooves callers to check the return values of functions.  If a function returns `dif_infalliable_t` I would expect them to check its value.  The cycle of "returning a meaningless value and then checking the meaningless value" strikes me as a high amount of overhead for not writing `void`.

File: sw/device/lib/dif/README.md (hunk 7 above)
Comment (gkelly) [reply sw/device/lib/dif/README.md]: I think it's more hassle to construct every output type if you're only interested in one of the values.

File: sw/device/lib/dif/README.md (hunk 2 above)
Comment (gkelly) [reply sw/device/lib/dif/README.md]: I believe in the defensive benefit of default here.

File: sw/device/lib/dif/README.md (hunk 10 above)
Comment (gkelly) [reply sw/device/lib/dif/README.md]: I think it's not worth opening the implementation details, given that we have provided One True Way of accessing MMIO regions already.

File: sw/device/lib/dif/README.md (hunk 7 above)
Comment (lenary) [reply sw/device/lib/dif/README.md]: We already have optional out-parameters, and they seem to work ok.

**Comment by lenary** [comment]
I think I've incorporated all the feedback so far. Thank you everyone for your suggestions and comments.

I think the next stage is to turn this into a real Pull Request, and to squash all the commits together for that. I will leave at least 24 hours before doing so, so people can comment on the most recent changes more easily.

**Comment by mcy** [comment]
@lenary A last minute suggestion.

I was reading the standard for unrelated reasons, and I was reminded that C has this syntax:
```c
void foo(int bar[static kLen]);
```
The `static` in the array declarator requests that the compiler make a best-effort attempt to check that the passed-in array has at least that length. `[static 1]` is, in effect, a compiler hint for "non null".
https://godbolt.org/z/06Qgnu

This has two downsides:
- Obviously, the compiler cannot catch most things. passing in any pointer value will always work, so it's not a replacement for null checks.
- This is not valid C++, and it would be rather nice if all our DIF headers were valid C++, since our tests want to include them in C++ files.

If we decide that this would be a nice bit of documentation that the compiler can sort-of enforce, we can add a macro
```c
#ifndef __cplusplus
  #define nonnull [static 1]
#else
  #define nonnull []
#endif
```
somewhere and write our DIF functions as
```c
void dif_my_thing_do(const dif_my_thing_t my_thing nonnull);
```

Thoughts? Feel free to shoot this down as overkill but I figured I'd mention it.

**Comment by lenary** [comment]
> @lenary A last minute suggestion.
> 
> I was reading the standard for unrelated reasons, and I was reminded that C has this syntax:
> 
> ```c
> void foo(int bar[static kLen]);
> ```
> 
> The `static` in the array declarator requests that the compiler make a best-effort attempt to check that the passed-in array has at least that length. `[static 1]` is, in effect, a compiler hint for "non null".
> https://godbolt.org/z/06Qgnu
> 
> This has two downsides:
> 
> * Obviously, the compiler cannot catch most things. passing in any pointer value will always work, so it's not a replacement for null checks.
> * This is not valid C++, and it would be rather nice if all our DIF headers were valid C++, since our tests want to include them in C++ files.
> 
> If we decide that this would be a nice bit of documentation that the compiler can sort-of enforce, we can add a macro
> 
> ```c
> #ifndef __cplusplus
>   #define nonnull [static 1]
> #else
>   #define nonnull []
> #endif
> ```
> 
> somewhere and write our DIF functions as
> 
> ```c
> void dif_my_thing_do(const dif_my_thing_t my_thing nonnull);
> ```
> 
> Thoughts? Feel free to shoot this down as overkill but I figured I'd mention it.

This is up for removal from C2x, because they totally forgot to figure it into C type compatibility rules, so `int foo(int bar[static 5])` is compatible with `int foo(int bar[])`, which of course is a bad thing. I am not willing to condone the use of static in this manner, much though I would like it to work :) 

In fact, let's try to avoid any of C11 that has an open defect report about it, as much as we can. Sadly, there is a DR about `volatile` that we cannot really avoid in order to access memory mapped registers.

**Comment by mcy** [comment]
I did not know about that DR. That sucks. Thanks for pointing that out.

> Sadly, there is a DR about volatile that we cannot really avoid in order to access memory mapped registers.
Yep, sounds about right. Technically, we could use `asm` if you really want avoid defect reports. =P

File: sw/device/lib/dif/README.md [hunk 11]
```diff
@@ -12,3 +12,74 @@ libraries.
...
+The guidelines below apply to writing DIFs, and code should be written in a
+similar style to the existing DIF libraries in this directory.
+
+### DIF Library Guidance
+
+* DIF libraries must be written in C.
+* DIF libraries can only depend on the following headers (and their associated
+  libraries) from the `sw/device/lib/base` directory:
+  * `sw/device/lib/base/mmio.h`
+  * `sw/device/lib/base/memory.h`
```
Comment (lenary) [inline sw/device/lib/dif/README.md]: These have been specifically enumerated because of #1200.

**Comment by lenary** [comment]
I will squash this commit today, with a view to merging by the end of the week.

Any last comments?

**Comment by imphil** [comment]
Ready for merging @lenary?
